### PR TITLE
Fix `prepare_alg` for `SplitODEProblem`s

### PR DESCRIPTION
**Pull request by vpuri3**
fixes https://github.com/SciML/OrdinaryDiffEq.jl/issues/1642

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by vpuri3** [comment]
@ChrisRackauckas merge?

**Comment by vpuri3** [comment]
last commit should fix https://github.com/SciML/LinearSolve.jl/issues/110. running CI

**Comment by ChrisRackauckas** [comment]
Needs a test.

**Comment by vpuri3** [comment]
Looks like there's another bug in `prepare_alg`

On master branch
```julia
using OrdinaryDiffEq, LinearAlgebra
n = 8
dt = 1//16
u0 = ones(n)
alg = SBDF3()
tspan = (0.0,1.0)

M1 = 2ones(n) |> Diagonal |> Array
M2 = 2ones(n) |> Diagonal |> Array

f1 = M1 |> DiffEqArrayOperator
f2 = M2 |> DiffEqArrayOperator
prob = SplitODEProblem(f1,f2,u0,tspan)

sol  = solve(prob, alg, dt=dt) # fails
@show sol.retcode

sol  = DiffEqBase.__solve(prob, alg, dt=dt) # works
@show sol.retcode

_alg = DiffEqBase.prepare_alg(alg,prob.u0,prob.p,prob)
sol = DiffEqBase.__solve(prob, _alg, dt=dt) # fails
```
```julia
julia> include("a.jl")
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol.retcode = :ConvergenceFailure
sol.retcode = :Success
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol.retcode = :ConvergenceFailure
```

**Comment by vpuri3** [comment]
So `prepare_alg` is setting `alg.linsolve=GenericFactorization`, and we have
```julia
alg = SBDF3()
sol  = DiffEqBase.__solve(prob, alg, dt=dt)
@show sol.retcode

alg = SBDF3(linsolve=GenericFactorization())
sol  = DiffEqBase.__solve(prob, alg, dt=dt)
@show sol.retcode
```
```julia
sol.retcode = :Success
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol.retcode = :ConvergenceFailure
```

**Comment by vpuri3** [comment]
The working case is defaulting to `Krylov.GMRES`
```julia
julia> ig = SciMLBase.init(prob, SBDF3(), dt=dt);

julia> ig.cache.nlsolver.cache.linsolve.cacheval |> typeof
Krylov.GmresSolver{Float64, Vector{Float64}}
```
need to figure out why generic factorization is failing.

**Comment by vpuri3** [comment]
```julia
julia> alg1 = SBDF3(linsolve=GenericFactorization());

julia> ig = SciMLBase.init(prob, alg1, dt=dt);

julia> ig.cache.nlsolver.cache.linsolve.A._concrete_form == ig.cache.nlsolver.cache.linsolve.cacheval
true

```

**Comment by ChrisRackauckas** [comment]
Test against the KenCarps first.

**Comment by vpuri3** [comment]
same behaviour
```julia
using OrdinaryDiffEq, LinearAlgebra, LinearSolve
n = 8
dt = 1/16
u0 = ones(n)
tspan = (0.0,1.0)

M1 = 2ones(n) |> Diagonal #|> Array
M2 = 2ones(n) |> Diagonal #|> Array

f1 = M1 |> DiffEqArrayOperator
f2 = M2 |> DiffEqArrayOperator
prob = SplitODEProblem(f1,f2,u0,tspan)

for algname in (
                :SBDF2,
                :SBDF3,
                :KenCarp47,
               )
    println("##########################################################")
    println("Testing out $algname")
    println("##########################################################")

    alg0 = @eval $algname()
    alg1 = @eval $algname(linsolve=GenericFactorization())

    kwargs = (dt=dt,)

    # OrdinaryDiffEq.prepare_alg() sets
    # alg.linsovle=GenericFactorization()
    @show sol  =              solve(prob, alg ; kwargs...).retcode # fails
    @show sol0 = DiffEqBase.__solve(prob, alg0; kwargs...).retcode # passes
    @show sol1 = DiffEqBase.__solve(prob, alg1; kwargs...).retcode # fails
end
```

```julia
julia> include("a.jl")
┌ Warning: Assignment to `kwargs` in soft scope is ambiguous because a global variable by the same name exists: `kwargs` will be treated as a new local. Disambiguate by using `local kwargs` to suppress this warning or `global kwargs` to assign to the existing global variable.
└ @ ~/.julia/dev/OrdinaryDiffEq.jl/test/interface/a.jl:27
##########################################################
Testing out SBDF2
##########################################################
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol = (solve(prob, alg; kwargs...)).retcode = :ConvergenceFailure
sol0 = (DiffEqBase.__solve(prob, alg0; kwargs...)).retcode = :Success
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol1 = (DiffEqBase.__solve(prob, alg1; kwargs...)).retcode = :ConvergenceFailure
##########################################################
Testing out SBDF3
##########################################################
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol = (solve(prob, alg; kwargs...)).retcode = :ConvergenceFailure
sol0 = (DiffEqBase.__solve(prob, alg0; kwargs...)).retcode = :Success
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol1 = (DiffEqBase.__solve(prob, alg1; kwargs...)).retcode = :ConvergenceFailure
##########################################################
Testing out KenCarp47
##########################################################
┌ Warning: Newton steps could not converge and algorithm is not adaptive. Use a lower dt.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:357
sol = (solve(prob, alg; kwargs...)).retcode = :ConvergenceFailure
sol0 = (DiffEqBase.__solve(prob, alg0; kwargs...)).retcode = :Success
┌ Warning: dt <= dtmin. Aborting. There is either an error in your model specification or the true solution is unstable.
└ @ SciMLBase ~/.julia/packages/SciMLBase/GW7GW/src/integrator_interface.jl:345
sol1 = (DiffEqBase.__solve(prob, alg1; kwargs...)).retcode = :DtLessThanMin
```

**Comment by vpuri3** [comment]
ok here's where im at

```julia
using OrdinaryDiffEq, LinearAlgebra, LinearSolve
n = 8 
dt = 1/16
u0 = ones(n)
tspan = (0.0,1.0)
  
M1 = 2ones(n) |> Diagonal |> Array
M2 = 2ones(n) |> Diagonal |> Array

f1 = M1 |> DiffEqArrayOperator
f2 = M2 |> DiffEqArrayOperator
prob = SplitODEProblem(f1,f2,u0,tspan)

alg0 = KenCarp47()                                # passing case
alg1 = KenCarp47(linsolve=GenericFactorization()) # failing case (factorizations are failing)
  
ig0 = SciMLBase.init(prob, alg0; dt=dt)
ig1 = SciMLBase.init(prob, alg1; dt=dt)

lincache0 = ig0.cache.nlsolver.cache.linsolve
lincache1 = ig1.cache.nlsolver.cache.linsolve

# check operators
W0 = lincache0.A
W1 = lincache1.A

for field in W0 |> typeof |> fieldnames
    isequal = getfield(W0, field) == getfield(W1, field)
    println("field $field is $isequal")
end

@show W0._concrete_form == W1._concrete_form # true
@show W0._func_cache    == W1._func_cache    # true

# check operator application
b = ones(n)
@show W0 * b == W1 * b                             # true
@show mul!(rand(n), W0, b) == mul!(rand(n), W1, b) # true
@show W0 \ b == W1 \ b                             # true

# check solve
lincache0.b .= 1.0
lincache1.b .= 1.0

solve(lincache0)
solve(lincache1)

@show lincache0.u == lincache1.u # false

@show lincache0.u 
@show lincache1.u
```
```julia
julia> include("a.jl")
field mass_matrix is true
field gamma is true
field J is true
field transform is true
field _func_cache is true
field _concrete_form is true
field jacvec is true
W0._concrete_form == W1._concrete_form = true
W0._func_cache == W1._func_cache = true
W0 * b == W1 * b = true
mul!(rand(n), W0, b) == mul!(rand(n), W1, b) = true
W0 \ b == W1 \ b = true
lincache0.u == lincache1.u = false
(lincache0.u)' = [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0]
(lincache1.u)' = [-1.1428571428571428 -1.1428571428571428 -1.1428571428571428 -1.1428571428571428 -1.1428571428571428 -1.1428571428571428 -1.1428571428571428 -1.1428571428571428]
```
because `solve(lincache1)` solves `Au=b` with a factorization method, it does not see the preconditioners `Pl, Pr` with diagonal weights which are `0` since we haven't hit `OrdinaryDiffEq.initialize!(nlsolver, integrator)` yet. This not accounting for the diagonal preconditioner is causing either the system or the error check to misbehave.

**Comment by vpuri3** [comment]
https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/src/nlsolve/utils.jl#L151-L153

**Comment by vpuri3** [comment]
@ChrisRackauckas

File: test/interface/a.jl [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+@show W0._concrete_form == W1._concrete_form # true
+@show W0._func_cache    == W1._func_cache    # true
+
+# check operator application
+b = ones(n)
+@show W0 * b == W1 * b                             # true
+@show mul!(rand(n), W0, b) == mul!(rand(n), W1, b) # true
+@show W0 \ b == W1 \ b                             # true
+
```
Comment (ChrisRackauckas) [inline test/interface/a.jl:65]: These should be tests

File: test/interface/a.jl [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+W0 = lincache0.A
+W1 = lincache1.A
+
+for field in W0 |> typeof |> fieldnames
+    isequal = getfield(W0, field) == getfield(W1, field)
+    println("field $field is $isequal")
+end
+
+@show W0._concrete_form == W1._concrete_form # true
+@show W0._func_cache    == W1._func_cache    # true
```
Comment (ChrisRackauckas) [inline test/interface/a.jl:58]: tests

File: test/interface/a.jl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+@show W0 \ b == W1 \ b                             # true
+
+# check solve
+lincache0.b .= 1.0
+lincache1.b .= 1.0
+
+solve(lincache0)
+solve(lincache1)
+
+@show lincache0.u == lincache1.u # false
```
Comment (ChrisRackauckas) [inline test/interface/a.jl:73]: test broken, and a working test

Comment (ChrisRackauckas) [reply test/interface/a.jl:73]: For the working test, use a non-operator reference

**Comment by vpuri3** [comment]
gonna try to debug this a little more with `LinearSolve.LinearSolveFunction()`

**Comment by vpuri3** [comment]
@ChrisRackauckas this is a good workaround
```julia
function linsolve(A,b,u,p,newA,Pl,Pr,solverdata;kwargs...)
    # avoid preconditioner monkeybusiness
    prob = LinearProblem(A,b;u0=u)
    solve(prob, nothing)
    return u
end

alg = KenCarp47(linsolve=LinearSolveFunction(linsolve))
@test solve(prob, alg).retcode == :Success # works
```

**Comment by vpuri3** [comment]
@ChrisRackauckas approve workflows again?

**Comment by ChrisRackauckas** [comment]
>@ChrisRackauckas this is a good workaround

That's not a real solution though.

**Comment by vpuri3** [comment]
it has to do with the weight vector and default preconditioner. idk where to look to debug

> because `solve(lincache1)` solves `Au=b` with a factorization method, it does not see the preconditioners `Pl, Pr` with diagonal weights which are `0` since we haven't hit `OrdinaryDiffEq.initialize!(nlsolver, integrator)` yet. This not accounting for the default preconditioner is causing either the system or the error check to misbehave in the `linsolve=factorization` cases.

**Comment by vpuri3** [comment]
@ChrisRackauckas do you know where to look? if not, can we merge with the one fix, and tests?

File: test/runtests.jl [hunk 4]
```diff
@@ -15,8 +15,10 @@ end
 
 @time begin
 
+
 if GROUP == "All" || GROUP == "InterfaceI" || GROUP == "Interface"
   #@time @safetestset "Linear Solver Tests" begin include("interface/linear_solver_test.jl") end
+  #@time @safetestset "Linear Solver Split ODE Tests" begin include("interface/linear_solver_split_ode_test.jl") end
```
Comment (ChrisRackauckas) [inline test/runtests.jl:21]: remove don't comment out.

Comment (vpuri3) [reply test/runtests.jl]: done

File: test/runtests.jl [hunk 5]
```diff
@@ -15,6 +15,7 @@ end
 
 @time begin
 
+
 if GROUP == "All" || GROUP == "InterfaceI" || GROUP == "Interface"
   #@time @safetestset "Linear Solver Tests" begin include("interface/linear_solver_test.jl") end
```
Comment (ChrisRackauckas) [inline test/runtests.jl:20]:
```suggestion
```

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
It's at least an improvement so I'll merge, but the full fix will take some time I don't have right now. I am not quite sure what it is without digging in.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
Test failures.

**Comment by vpuri3** [comment]
failing with 1.6
```julia
julia> versioninfo()
Julia Version 1.6.6
Commit b8708f954a (2022-03-28 07:17 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, broadwell)
Environment:
  JULIA_NUM_PRECOMPILE_TASKS = 4
  JULIA_DEPOT_PATH = /Users/vp/.julia
  JULIA_NUM_THREADS = 4
```


```julia
[vp@MBP ~]:/Applications/Julia-1.6.app/Contents/Resources/julia/bin/julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.6 (2022-03-28)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Test

julia> @test_warn "hi" @warn "hi hello"
┌ Warning: hi hello
└ @ Main REPL[2]:1
Test Failed at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:671
  Expression: contains_warn(read(fname, String), $(Expr(:escape, "hi")))
ERROR: There was an error during testing
```

**Comment by devmotion** [comment]
That's expected. From the docs of `@test_warn` (https://docs.julialang.org/en/v1/stdlib/Test/#Test.@test_warn):

> Note: Warnings generated by @warn cannot be tested with this macro. Use @test_logs instead.

**Comment by vpuri3** [comment]
oops thanks for correcting that dumb mistake @devmotion . @ChrisRackauckas approve workflow again?

**Comment by vpuri3** [comment]
weird the last check has been pending for the last 6 hours

**Comment by ChrisRackauckas** [comment]
Our GPU machines appear to be down @staticfloat ?

**Comment by staticfloat** [comment]
I don't manage those, that's all Tim.

@maleadt Looks like (almost) all the gpuci machines have gone down?

**Comment by maleadt** [comment]
Yeah, I can't even reach the server. I'll ping Valentin on the Slack community-ci channel.

**Pull request merged**